### PR TITLE
Reduces number of remove VIP operations in svcs

### PR DESCRIPTION
**Pull request by trozet**
We see at scale that we are calling remove VIP way too many times due to
the fact that we have to ensure that the VIP does not exist on another
load balancer when we update it. We were blindly doing that for idling
and the cluster vs worker loadbalancers.

This patch addresses it by using a cache inside service tracker to track
which type of load balancer the VIP is currently on. If the entry
doesn't exist, we brute force remove as before, but if it does exist we
try to avoid unnecessary nbctl commands to remove.

Note, using go-ovn or libovsdb bindings would be a better solution here,
but that implementation can come later as it will require a much bigger
change to the project.

Signed-off-by: Tim Rozet <trozet@redhat.com>

### CODE REVIEW

**Comment by trozet** [comment]
@astoycos FYI we will probably need to do this for the new load balancer you are adding too. That will add even more remove commands.

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 1]
```diff
@@ -271,29 +274,68 @@ func (c *Controller) syncServices(key string) error {
...
+		// create a temporary map to convert sets.String to map for passing to functions
+		tmpVIPs := make(map[string]string)
+		for v := range vips {
+			delete(vipsTracked, v)
+			tmpVIPs[v] = ""
+		}
+		needsCleanUp := false
+		for vipProtocol := range vips {
+			vip, _ := splitVirtualIPKey(vipProtocol)
+			if c.serviceTracker.getLoadBalancer(name, namespace, vip) != loadbalancer.IdlingLoadBalancer {
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:293]: Can we merge the two range loops and add to tmpVIPs only if this if matches?
So tmpVIPs contains only those vip that need to be removed from non idling loadbalancers.
With that, you can get rid of needsCleanUp because the map will be empty if there are no matches for the if

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 2]
```diff
@@ -271,29 +274,68 @@ func (c *Controller) syncServices(key string) error {
...
 		return nil
 	}
 
-	err = deleteVIPsFromIdlingBalancer(vips, name, namespace)
-	if err != nil {
-		c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToDeleteOVNLoadBalancer",
-			"Error trying to delete the idling OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
-		return err
+	needsCleanUp := false
+	tmpVIPs := make(map[string]string)
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:321]: can you rename tmpVIPs in "toRemoveFromIdling" or something like that?

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 3]
```diff
@@ -271,29 +274,68 @@ func (c *Controller) syncServices(key string) error {
...
 		err = c.addServiceToIdlingBalancer(vips, service)
 		if err != nil {
 			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToAddToIdlingBalancer",
 				"Error trying to add to Idling LoadBalancer for Service %s/%s: %v", name, namespace, err)
 			return err
 		}
-		vipsTracked = vipsTracked.Delete(vips.List()...)
-		err = deleteVIPsFromNonIdlingOVNBalancers(vips, name, namespace)
+		// create a temporary map to convert sets.String to map for passing to functions
+		tmpVIPs := make(map[string]string)
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:285]: can you rename tmpVIPs in `toRemoveFromNonIdling` ?

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 4]
```diff
@@ -271,29 +274,68 @@ func (c *Controller) syncServices(key string) error {
...
+	tmpVIPs := make(map[string]string)
+	for vipProtocol := range vips {
+		vip, _ := splitVirtualIPKey(vipProtocol)
+		foundLb := c.serviceTracker.getLoadBalancer(name, namespace, vip)
+		// if vip was on an idling load balancer and we get here, we know we need to clean up
+		// if the value is empty (unknown what load balancer the vip may have existed on) also need to clean up
+		if len(foundLb) == 0 || foundLb == loadbalancer.IdlingLoadBalancer {
+			needsCleanUp = true
+		}
+		tmpVIPs[vipProtocol] = ""
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:330]: Same as above, can you put this under the condition? This will make it possible to get rid of `needsCleanUp` (and will save some unnecessary calls to deleteVIPsFromIdlingBalancer

**Comment by aojea** [comment]
The fundamental problem here is the migration from VIPs from one LB to another LB, since we don't know where the VIP was, we just do the "delete VIP in LB1/create VIP in LB2" to emulate a "moveVIPIfExistFromLB1toLB2".

I was thinking that the nbctl command was caching operations and will detect the deletes as noops. but it doesn't seem so and it seems that there are a lot of "execs" causing high cpu consumption with a lot of services.

I vaguely remember that we also took this pattern for other features, switching the VIP on different LB, that is really a hack to deal with some architectural/OVN limitations we did. This approach seems to perpetuate this hack, making the code still more confusing and complicated.

If we are going to keep using this pattern I rather invest in having a way of implementing "moveVIPIfExistFromLB1toLB2()"  that we can reuse and is efficient, instead of start to overload current structures to workaround this limitation.

It seems go-bindings can't help us here on implementing the transaction directly or does it?
I was wondering if it can help at least to know where the VIP is/was, my understanding is that go-bindings maintain a local cache of the ovsdb, can we do a `if isVIPinThisLb1(vip) then deleteVIP(lb1) and createVIP(lb2) `? 
that will save undesired deleted consulting the go-bindings local cache and only delete if necessary.

File: go-controller/pkg/ovn/controller/services/service_tracker.go [hunk 5]
```diff
@@ -27,23 +26,24 @@ func splitVirtualIPKey(key string) (string, v1.Protocol) {
...
+	// holds a map of service mapping to map of vips and loadbalancers
+	virtualIPByService map[string]map[string]string
 	hadEndpoints       map[string]bool
 }
 
 // newServiceTracker creates and initializes a new serviceTracker.
 func newServiceTracker() *serviceTracker {
 	return &serviceTracker{
-		virtualIPByService: map[string]sets.String{},
+		virtualIPByService: map[string]map[string]string{},
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/service_tracker.go:40]: This double map is quite hard to follow.
I'd have the inner one described by a type with methods (which can still be baked by a map).

**Comment by trozet** [comment]
@aojea yeah bindings will eliminate the need for this because we would have a local cache we could just check and not do any operations if the VIP is not stale on another LB. I started down that path first, but then realized the go-ovn bindings are missing the functionality I needed. We are going to be moving from go-ovn bindings to libovsdb sometime in the future (I hope soon), so imo this is something we have to live with for a short amount of time. I view this solution as temporary until we can use an ovn cache.

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 6]
```diff
@@ -271,29 +274,68 @@ func (c *Controller) syncServices(key string) error {
...
+			if err != nil {
+				c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToDeleteOVNLoadBalancer",
+					"Error trying to delete the OVN LoadBalancer while setting up Idling for Service %s/%s: %v", name, namespace, err)
+				return err
+			}
+		}
+		// at this point we have processed all vips we've found in the service
+		// so the remaining ones that we had in the vipsTracked variable should be deleted
+		// We remove them from OVN and from the tracker
+		err = deleteVIPsFromAllOVNBalancers(vipsTracked, name, namespace)
```
Comment (aojea) [inline go-controller/pkg/ovn/controller/services/services_controller.go:309]: did we processed all vips here?
I think this was an `		err = deleteVIPsFromNonIdlingOVNBalancers(vips, name, namespace)` before

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 3 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:285]: done

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 1 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:293]: good idea, will do

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 6 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:309]: the delete from non idling still happens above, but if we get here we are going to return out on line 317, and we were missing cleaning up OVN, but we were deleting the vips from the tracker. I broke this out into its own commit to make this less confusing (and ensure we dont revert this fix in the future)

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 4 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:330]: this is what 12 am code looks like :)

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 2 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:321]: done

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 7]
```diff
@@ -234,8 +234,11 @@ func (c *Controller) syncServices(key string) error {
 	// These are the VIPs (ClusterIP:Port) that we have seen so far
 	// If the Service has updated the VIPs (has changed the Ports)
 	// and some were removed we have to delete those
-	// We need to create a NewString set to not mutate the service tracker VIPs
-	vipsTracked := sets.NewString().Union(c.serviceTracker.getService(name, namespace))
+	// We need to create a map to not mutate the service tracker VIPs
+	vipsTracked := make(map[string]string)
+	for k, v := range c.serviceTracker.getService(name, namespace) {
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:239]: I think vipsTracked can be a set.String now?
I checked in all the places we use, we only access the key.

This would result in changing the methods argument as well.

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 8]
```diff
@@ -271,25 +274,51 @@ func (c *Controller) syncServices(key string) error {
...
 		err = c.addServiceToIdlingBalancer(vips, service)
 		if err != nil {
 			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToAddToIdlingBalancer",
 				"Error trying to add to Idling LoadBalancer for Service %s/%s: %v", name, namespace, err)
 			return err
 		}
-		vipsTracked = vipsTracked.Delete(vips.List()...)
-		err = deleteVIPsFromNonIdlingOVNBalancers(vips, name, namespace)
+		// create a temporary map to convert sets.String to map for passing to functions
+		toRemoveFromNonIdling := loadbalancersPerVip{}
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:285]: Same, this can be a set.

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 9]
```diff
@@ -271,25 +274,51 @@ func (c *Controller) syncServices(key string) error {
...
+				"Error trying to delete the OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+			return err
+		}
 		// Delete the Service VIP from the Service Tracker
 		c.serviceTracker.deleteServiceVIPs(name, namespace, vipsTracked)
 		return nil
 	}
 
-	err = deleteVIPsFromIdlingBalancer(vips, name, namespace)
+	toRemoveFromIdling := loadbalancersPerVip{}
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:312]: Same

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 7 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:239]: will do

File: go-controller/pkg/ovn/controller/services/repair.go [hunk 10]
```diff
@@ -70,15 +70,35 @@ func (r *Repair) runOnce() error {
 				lbUUID, err := gateway.GetGatewayLoadBalancer(gatewayRouter, p)
 				if err != nil {
 					if err != gateway.OVNGatewayLBIsEmpty {
-						klog.V(5).Infof("Failed to get OVN GR load balancer for protocol %s, err: %v", p, err)
+						klog.V(5).Infof("Failed to get OVN GR: %s load balancer for protocol %s, err: %v",
```
[the remark this answers is not in the document]
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/repair.go:73]: im going to add this to another PR where I have a few random fixes to the services code like this

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 11]
```diff
@@ -271,25 +271,52 @@ func (c *Controller) syncServices(key string) error {
...
+		// addServiceToIdlingBalancer adds the vips to service tracker
 		err = c.addServiceToIdlingBalancer(vips, service)
 		if err != nil {
 			c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToAddToIdlingBalancer",
 				"Error trying to add to Idling LoadBalancer for Service %s/%s: %v", name, namespace, err)
 			return err
 		}
-		vipsTracked = vipsTracked.Delete(vips.List()...)
-		err = deleteVIPsFromNonIdlingOVNBalancers(vips, name, namespace)
+		// create a temporary map to convert sets.String to map for passing to functions
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:281]: nit: not a map anymore

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 12]
```diff
@@ -271,25 +271,52 @@ func (c *Controller) syncServices(key string) error {
...
+		toRemoveFromNonIdling := sets.NewString()
+		for vipProtocol := range vips {
+			vip, _ := splitVirtualIPKey(vipProtocol)
+			if c.serviceTracker.getLoadBalancer(name, namespace, vip) != loadbalancer.IdlingLoadBalancer {
+				toRemoveFromNonIdling.Insert(vip)
+			}
+			delete(vipsTracked, vipProtocol)
```
Comment (fedepaol) [inline go-controller/pkg/ovn/controller/services/services_controller.go:288]: should we use the sets Delete method here?

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 11 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:281]: will fix

File: go-controller/pkg/ovn/controller/services/services_controller.go (hunk 12 above)
Comment (trozet) [reply go-controller/pkg/ovn/controller/services/services_controller.go:288]: yep, will fix

File: go-controller/pkg/ovn/controller/services/services_controller.go [hunk 13]
```diff
@@ -271,25 +271,51 @@ func (c *Controller) syncServices(key string) error {
...
 				"Error trying to add to Idling LoadBalancer for Service %s/%s: %v", name, namespace, err)
 			return err
 		}
-		vipsTracked = vipsTracked.Delete(vips.List()...)
-		err = deleteVIPsFromNonIdlingOVNBalancers(vips, name, namespace)
+		toRemoveFromNonIdling := sets.NewString()
+		for vipProtocol := range vips {
+			vip, _ := splitVirtualIPKey(vipProtocol)
+			if c.serviceTracker.getLoadBalancer(name, namespace, vip) != loadbalancer.IdlingLoadBalancer {
+				toRemoveFromNonIdling.Insert(vip)
```
Comment (trozet) [inline go-controller/pkg/ovn/controller/services/services_controller.go:285]: @fedepaol helped me figure out this is causing a segfault cause I'm passing the vip here and not vipProtocol. That is why the unidling tests are failing. will fix.

**Comment by trozet** [comment]
As long as this passes e2e unidilng tests I think we should be good to go. Tested it downstream in scale lab and seems good.
